File: examples/boardgame_g2d.py
#!/usr/bin/env python3

import g2d
from boardgame import BoardGame
from time import time

W, H = 40, 40
LONG_PRESS = 0.5

class BoardGameGui:
    def __init__(self, g: BoardGame):
        self._game = g
        self._downtime = 0
        self.update_buttons()

    def tick(self):
        if g2d.key_pressed("LeftButton"):
            self._downtime = time()
        elif g2d.key_released("LeftButton"):
            pos = g2d.mouse_position()
            x, y = pos[0] // W, pos[1] // H
            if time() - self._downtime > LONG_PRESS:
                self._game.flag_at(x, y)
            else:
                self._game.play_at(x, y)
            self.update_buttons()

    def update_buttons(self):
        g2d.clear_canvas()
        g2d.set_color((0, 0, 0))
        rows, cols = self._game.rows(), self._game.cols()
        for y in range(1, rows):
            g2d.draw_line((0, y * H), (cols * W, y * H))
        for x in range(1, cols):
            g2d.draw_line((x * W, 0), (x * W, rows * H))
        for y in range(rows):
            for x in range(cols):
                value = self._game.value_at(x, y)
                center = x * W + W//2, y * H + H//2
                g2d.draw_text_centered(value, center, H//2)
        g2d.update_canvas()
        if self._game.finished():
            g2d.alert(self._game.message())
            g2d.close_canvas()

def gui_play(game: BoardGame):
    g2d.init_canvas((game.cols() * W, game.rows() * H))
    ui = BoardGameGui(game)
    g2d.main_loop(ui.tick)

File: go/boardgame/boardgame_g2d.go
package boardgame

import . "g2d"
import "time"

var W, H int = 40, 40
var LONG_PRESS float64 = 0.5

type BoardGameGui struct {
    game BoardGame
    downtime time.Time
}

func NewBoardGameGui(game BoardGame) *BoardGameGui {
    gui := &BoardGameGui{game, time.Time{}}
    InitCanvas(Size{game.Cols() * W, game.Rows() * H})
    gui.Update()
    HandleKeys(gui.keydown, gui.keyup)
    return gui
}

func (gui *BoardGameGui) MainLoop() {
    MainLoop(nil, 1000/30)
}

func (gui *BoardGameGui) keydown(code string, pos Point) {
    if code == "LeftButton" {
        gui.downtime = time.Now()
    }
}

func (gui *BoardGameGui) keyup(code string, pos Point) {
    if code == "LeftButton" {
        if time.Now().Sub(gui.downtime).Seconds() > LONG_PRESS {
            gui.game.FlagAt(pos.X, pos.Y)
        } else {
            gui.game.PlayAt(pos.X, pos.Y)
            gui.Update()
        }
    }
}

func (gui *BoardGameGui) Update() {
    FillCanvas(Color{255, 255, 255})
    rows, cols := gui.game.Rows(), gui.game.Cols()
    for y := 1; y < rows; y++ {
        DrawLine(Color{0, 0, 0}, Point{0, y * H}, Point{cols * W, y * H})
    }
    for x := 1; x < cols; x++ {
        DrawLine(Color{0, 0, 0}, Point{x * W, 0}, Point{x * W, rows * H})
    }
    for y := 0; y < rows; y++ {
        for x := 0; x < cols; x++ {
            DrawTextCentered(gui.game.GetVal(x, y), Color{0, 0, 0},
                Point{x * W + W/2, y * H + H/2}, H/2)
        }
    }
    UpdateCanvas()
    if gui.game.Finished() {
        Alert(gui.game.Message())
        // g2d.draw_text_centered(self._game.message(), (0, 0, 255),
        //     (cols * W/2, rows * H/2), H/2)
        Exit()
    }
}
